### ISSUE TITLE
Missing commits after merge

### ISSUE THREAD

**Issue by schmurfy** [concourse/git-resource#340]
For my use case I get the resource the main project depend on on each PR with the idea to check that these dependencies have been merge in master and are not on a branch.
My idea would work if not for a strange behaviour from the git resource which may be due to the  `--single-branch` flag, here is my repository history:
<img width="89" alt="screenshot_617" src="https://user-images.githubusercontent.com/2983/99267118-d48e6400-2823-11eb-8953-34bc5dbec735.png">

And here is what the resource sees:
![screenshot_618](https://user-images.githubusercontent.com/2983/99267165-e243e980-2823-11eb-93df-742d30234a3c.png)

It only lists the commits which are directly on the master branch and in this case the merge commit but what I expected to see is what git log shows me:
```
commit 77f7ac58bee341f9325c850eab2a0791826918ed (HEAD -> master, origin/master)
Merge: 641cc5c 05a0b2a
Author: [readacted]
Date:   Thu Nov 12 20:50:56 2020 +0100


commit 05a0b2a2cd47fe5be0ec97b62256caa6e6130a2a (origin/feature/add-country-code)
Author: [readacted]
Date:   Thu Nov 12 18:28:27 2020 +0100


commit 81b8933e3ea9a3c15417db587792cfc66bc54c50
Author: [readacted]
Date:   Thu Nov 12 18:26:13 2020 +0100
```

Is there a way to get that without forking this resource ?

Edit:
I did a check by hijacking the resource check container and removing `--single-branch` does indeed fix my issue

**Comment by schmurfy** [comment]
I forked the resource to do some tests but I still fails to find the rights arguments for `git rev-list`, I had no idea it could be so hard to get all the last commits on a branch... 😢 

If anyone can point me toward the right direction it would be awesome :)

What I did in the check script:
- removed `--single-branch` and `--first-parent`
After that it mostly works but since it `${ref}~1..HEAD` as path it always found the commit even if it was not merged so I tried using `${ref}..HEAD` but now the results is really weird...
What I actually hope to do is to list the last N commits leading to the tip of the branch (`master` here) including those from merged branches.

PS: I have no problem forking this resource if I get the behaviour I want, although it might be nice to have the option directly here.

**Comment by vito** [comment]
Hmm, as I recall `--first-parent` is necessary for reasonable CI behavior in more complicated cases. Your example is a simple fast-forward, so it's technically OK to walk through all commits, but in a scenario where e.g. `foo` branched off from `master`, more commits were then made on `master`, and commits were made on `foo`, if the resource were to return all commits you'd have the changes on `master` "disappear" when CI runs the commits on `foo`, and then reappear when the merge occurs.

So using `--first-parent` ensures that changes are tested in the actual order that they effectively appeared on `master`. Resources represent a single linear stream of versions, which is only possible if you limit the results to a single branch. If it were to include the other commits it would technically be alternating between branches, effectively changing what's actually being tested each time.

Does that make sense? Sorry it's not really a solution, but maybe it'll help lead to other ideas. 🙂 (I also admit I don't 100% understand what you're trying to do.)

**Comment by schmurfy** [comment]
I will try to explain better my usecase but I agree it is pretty niche.

let's say we have multiple repositories:
- app1
- lib1
- lib2

for a specific feature I create a PR on `lib1` and another on `app1` which points to `lib1` pr branch, before concourse give the green light to let the `app1` pr be merged I want a way to check that the version of `lib1` and `lib2` imported comes from master and are not on an unmerged branch.

What I do is (using jsonnet) dynamically create a pipeline when a pr is open with the list of repositories and their sha1 this pr depends on and the last task in that pipeline get all these as resource, doing that this way have some advantages:
- the obvious is that the pipeline will be blocked if one or more dependencies have not been merged
- the bonus is that when they get merged the pipeline will continue and complete

For now I have something which seems to work with a modified git-resource but I understand well that this behaviour might not be needed by anyone else, this resource handle a lot of usecases already.
I use the forked git recourse only to check on these dependencies.

PS: I agree that this behaviour is not desirable for standard usage

**Comment by vito** [comment]
@schmurfy Interesting - almost sounds like you want something like `git log --first-parent --having-ancestor SHA` (don't get too excited, I made that second flag up). Basically, only consider the final merged commits, and filter them such that they have the given SHA as an ancestor. With this configuration the resource wouldn't return anything until `SHA` is merged, and at that point your pipeline would become unblocked and run with the merge commit.

Alternatively, you could add an extra job to your pipeline that just fails if the desired commit isn't present in the repo.

```
[lib1] -> {check-for-SHA} -> [lib1] -> {test-app1}
```

I'm going to close this for now as it's working as intended and it sounds like you're OK with using a fork. The above approach might also make sense. If there ends up being a straightforward way to do this with `git` I'd be happy to consider a PR, but yeah, this resource is already getting pretty unwieldy.
